design/bios-config.md

```diff
@@ -0,0 +1,163 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# bios-config
+
+## Status
+
+provisional
+
+## Table of Contents
+
+<!--ts-->
+   * [Title](#title)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+This document explains the usage of YAML attributes for BIOS configuration for the BareMetalHost of the BMO. These configurations will be applied to the host according to the vendor's BMC driver being used by Ironic.
+
+## Motivation
+
+We need this document to specify the typical usage of BIOS YAML attributes in the BareMetalHost YAML. 
+
+
+### Goals
+
+1. To agree on the format of the attributes that are to be supported as a starting point
+2. To keep the attribute naming vendor agnostic
+
+### Non-Goals
+
+1. To list every attribute for every vendor exhaustively
+
+## Proposal
+The proposal is to set/unset the BIOS configuration values using vendor drivers available for each BMC type. For this purpose, new YAML attributes are to be introduced in the BareMetalHost spec. 
+ 
+### Implementation Details/Notes/Constraints 
+All the BIOS related attributes or fields will come under the sub-section called ```bios``` in spec. The values for each attribute can be a boolean or a string. The vendor driver type to implement this configuration is not to be specified separetely as it is already known from the bmc sub-section.
+
+The proposed BMH looks like the following, given the three supported parameters
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+ name: bm0
+spec:
+ online: true
+ bmc:
+   address: idrac://192.168.122.1:6230/
+   credentialsName: bm0-bmc-secret
+ bootMACAddress: 52:54:00:b7:b2:6f	
+ bios:
```

Comment:
@stbenjam pointed out that for UEFI these options aren't part of the BIOS because BIOS isn't used. Maybe we should call these "firmware" settings instead to avoid confusion?

Reply:
```bios``` was used to conform more with the generally accepted notion of BIOS. Sure, we can call this ```firmware```, given that it will be clear enough for the user. What do you think?

Reply:
Yea, I would prefer `firmware` - UEFI != BIOS, but it might be too pedantic of a distinction for most people since everyone might just call it "bios"  anyway. I don't want to hold this up on bikeshedding either...

Maybe some ironic folks haver opinions about this - @juliakreger @dtantsur.

Reply:
That's precisely my concern as well @stbenjam

Reply:
Another point is that `bios` is basically irrelevant outside of the x86 world. I'd rather we be a bit future proof here and use a generic word for this (e.g. firmware)

Reply:
Alright, sure, I can rename this to "firmware". That might expose this section to a lot more 'potential information' like firmware versions and such, is that alright? because it might not mean just BIOS settings then.

Reply:
If we ever expose an API to update the firmware version, I would expect that to end up in this section, too. So yeah, I think that's OK. If it turns out to be not what we want, then we can come up with another name for a new section when we understand the distinction we're trying to make.

Reply:
That sounds good to me. I can add this point inside the spec above this, allowing us to document this thought for future reference. Let me send a patch for this.

Reply:
Thanks, I like it. One note, someone else suggested "systemFirmware" so as to differentiate it from any other kind. Although maybe it's a top-level item in the baremetal host CRD, the system is probably assumed.

Reply:
Great! Yes, the notion of system will be implied.